### PR TITLE
CP-29049: Fix list and abort behaviour and expand protocol tests

### LINKED ISSUE #130: Server does not respond with `NBD_REP_ACK` to `NBD_OPT_ABORT`
The [NBD protocol docs](https://github.com/NetworkBlockDevice/nbd/blob/master/doc/proto.md) says:
> NBD_OPT_ABORT (2)

> The client desires to abort the negotiation and terminate the session. The server MUST reply with NBD_REP_ACK.

However, this issue probably won't cause problems, because

> Previous versions of this document were unclear on whether the server should send a reply to NBD_OPT_ABORT. Therefore the client SHOULD gracefully handle the server closing the connection after receiving an NBD_OPT_ABORT without it sending a reply. Similarly the server SHOULD gracefully handle the client sending an NBD_OPT_ABORT and closing the connection without waiting for a reply.

### PR DESCRIPTION
**Pull request by gaborigloi**
Various fixes to make the client and server conform more to the NBD
protocol:
* Ensure Client.list terminates the option haggling by sending an abort.
  This change should not cause problems, as currently we don't seem to
  use Client.list anywhere.
* Fix the server's response to NBD_OPT_LIST - previously it did not send
  the lenghts to the client correctly, and therefore qemu-img and
  nbd-client could not list the exports.
* Check that the lengths the export names in the optional ~offer param
  passed to Server.connect are below the maximum limit allowed by the NBD
  protocol.
* Ensure the server responds to an abort with an ack. This change
  shouldn't cause issues either, because we gracefully handle the case
  without failing where the client disconnects without waiting for the
  ack.
* Clean up existing protocol unit tests, add more logging, and more
  tests for these corner cases.
* Add a new client<->server interaction test to check that NBD_OPT_LIST
  is implemented properly in both.

Fixes #130 #128 #102 

Signed-off-by: Gabor Igloi <gabor.igloi@citrix.com>

### CODE REVIEW

**Comment by gaborigloi** [comment]
I've checked, and after this change, if I specify the export name, it works with `qemu-img` and `nbd-client`:
```
~/s/nbd (fix_list_abort|✚1) $ _build/default/cli/main.exe serve --no-tls /tmp/d
main.exe: Server.connect: Starting fixed-newstyle negotiation
main.exe: Server.serve: Serving new client, read_only = false
main.exe: Server.serve: Received NBD_CMD_DISC, disconnecting
^C⏎
~/s/nbd (fix_list_abort|✚1) $ git diff
diff --git a/cli/main.ml b/cli/main.ml
index 8cfb25a..4755e75 100644
--- a/cli/main.ml
+++ b/cli/main.ml
@@ -183,6 +183,7 @@ module Impl = struct
              (fun clearchan ->
                 Server.with_connection
                   clearchan
+                  ~offer:["test"]
                   (fun _exportname svr ->
                      Nbd_lwt_unix.with_block filename
                        (Server.serve svr ~read_only:false (module Block))
~/s/nbd (fix_list_abort|✚1) $ _build/default/cli/main.exe serve --no-tls /tmp/d
main.exe: Server.connect: Starting fixed-newstyle negotiation
main.exe: main: Caught exception Nbd__Server.Client_requested_abort while handling connection
^C⏎
~/s/nbd (fix_list_abort|✚1) $ 
```

```
~/s/nbd (fix_list_abort|✚1) $ qemu-img convert nbd:0.0.0.0:10809:exportname=test -O qcow2 /tmp/out
~/s/nbd (fix_list_abort|✚1) $ nbd-client -l 0.0.0.0
Negotiation: ..
test
~/s/nbd (fix_list_abort|✚1) $ 
```
